### PR TITLE
Issue #9957: Fixing Unnecessary Parentheses false positives

### LINKED ISSUE #9957: UnnecessaryParentheses false positives in 8.42
This must be related to https://github.com/checkstyle/checkstyle/issues/5779 -- this behavior appeared in 8.42:

```bash
$ cat TestClass.java
public class TestClass {
    public static void main(String[] args) {
        boolean a = true;
        int b = 42;
        int c = 42;
        System.out.println(a == (b == c));
    }
}
```

```bash
$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
         <module name="UnnecessaryParentheses" />
    </module>
</module>
```

```bash
$ java -jar checkstyle-8.42-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] /mnt/c/Users/ipono/Downloads/__files/TestClass.java:6:33: Unnecessary parentheses around expression. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 1 errors.
```

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #9957: UnnecessaryParentheses false positives in 8.42
Diff Regression projects: https://gist.githubusercontent.com/Vyom-Yadav/91807e6244cff60698d9e33e32ba2d51/raw/eb228fa41edadc10eea264a914ea1411f0e52122/projects-to-test-on.properties
Diff Regression config: https://gist.githubusercontent.com/Vyom-Yadav/98dceb63a79f4833e85fff9b2e1464a6/raw/518edf2da1dd8da7ff8f5570c2ebb89671fcfcfb/my_checks.xml

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
@romani With most of the previous functionality dropped, the check now only checks unnecessary parentheses around unary and postfix operator. There were certain cases for which previous functionality, which checked unnecessary parentheses around conditional and relational operators. It was not possible for to make a check that used token types to check. All the checking for conditional and relational operators was dropped due to cases like-
```
boolean a = true;
int b = 42;
int c = 42;
int d = 32;
boolean bc = a==(c==d && d!=b)); // Cannot remove parentheses.
boolean ab = a!=(b==c); // cannot remove parentheses.
```
But still the checking for postfix and unary operators is there. I did not remove all the existing test cases which weren't showing violation now (may be useful in future), but if you want, I can remove them too.

**Comment by Vyom-Yadav** [comment]
@romani @rnveach @strkkk I tired to cover all the case giving false positives. Cases like-
```
int a = 5;
int b = 5;
boolean c = true;
if (a==(b <= 4 && c >= b)) {} 
if(a!=(b <= c)) {}
```
And further modification of these cases. I tried to keep most of the functionality of the check, if there are any cases you wish to check or I missed out please point them out so I can accommodate them too. I will add examples in doc and a note about certain limitation after you give a thumbs up to the changes and we make sure there are no more false positives.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 1]
```diff
@@ -39,10 +39,7 @@
  * int x = (y / 2 + 1); // parens around assignment rhs
  * for (int i = (0); i &lt; 10; i++) {  // parens around literal
  * t -= (z + 1);                     // parens around assignment rhs
- * boolean a = (x &gt; 7 &amp;&amp; y &gt; 5)      // parens around expression
- *             || z &lt; 9;
- * boolean b = (~a) &gt; -27            // parens around ~a
- *             &amp;&amp; (a-- &lt; 30);        // parens around expression
+ * boolean b = (~a) &gt; -27;            // parens around ~a
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:42]: Cases removed do not fall under the convention we said we would target. https://github.com/checkstyle/checkstyle/issues/9957#issuecomment-829744968

Why were these removed?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 2]
```diff
@@ -170,11 +167,6 @@
  *   .filter((s) -&gt; s.startsWith(&quot;c&quot;)) //violation
  *   .forEach(System.out::println);
  * int a = 10, b = 12, c = 15;
- * if ((a &gt;= 0 &amp;&amp; b &lt;= 9)            // violation, unnecessary parenthesis
- *          || (c &gt;= 5 &amp;&amp; b &lt;= 5)    // violation, unnecessary parenthesis
- *          || (c &gt;= 3 &amp;&amp; a &lt;= 7)) { // violation, unnecessary parenthesis
- *     return;
- * }
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:177]: Same.

File: src/xdocs/config_coding.xml [hunk 3]
```diff
@@ -7012,21 +7009,21 @@ int x = (a + b) + c;
                 ,<a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#TEXT_BLOCK_LITERAL_BEGIN">
                   TEXT_BLOCK_LITERAL_BEGIN</a>
                 ,<a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LAND">
-                  LAND</a>
+                LAND</a>
```
Comment (rnveach) [inline src/xdocs/config_coding.xml:7012]: All of these were not fixed. It may be better to revert these back and provide a fix in another PR.

**Review by rnveach — CHANGES_REQUESTED** [review]
We should add some text and examples to the xdoc that we will ignore these new conventions and why.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 2 above)
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:177]: Same as above.

**Review by rnveach — APPROVED** [review]

**Review by romani — APPROVED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 4]
```diff
@@ -105,6 +105,38 @@ private void UnaryAndPostfix() {
...
 
+    public void checkBooleanStatements() {
+        boolean a = true;
+        int b = 42;
+        int c = 42;
+        int d = 32;
+        if((b == c) == a
+                && (( // violation, unnecessary parenthesis, the latter one
+                                (b==c)==(d>=b)==a!=(c==d))
+                || (b<=c)!=a==(c>=d))) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:116]: your example are over complicated..... ok to keep it. 
but please add several simple examples that easy to consume by people.
please use expression from issue and please take few cases from diff report.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java (hunk 4 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java]: Done.

**Review by romani — APPROVED** [review]

**Pull request merged**